### ISSUE TITLE
ResultExpressions should have access to process variables

### ISSUE THREAD

**Issue by chDame** [camunda/connector-sdk#312]
Hi team,

I should be able to read process variables in the the result expression (like comments in bold in the following example) :
```
{
  status: response.body.status,
  comments: append(**comments**, {"comment":string(today())+" : Status "+response.body.status, "author": "system"})
}
```

But the result is NULL because the process variables are not available at evaluation: https://github.com/camunda/connector-sdk/blob/0c32b5501bbb5632727f5253d8a13b502161ddd0/runtime-util/src/main/java/io/camunda/connector/runtime/util/ConnectorHelper.java#L58

__Steps to Reproduce__

1. start a process with a comments variable that is a list (empty or not)
2. call a rest endpoint with the rest connector that returns the payload {"status":"youpi"}
3. use the resultExpression provided in this bug report

__Expected Behavior__

I should be able to enrich existing process variables with incoming results

__Environment__
Camunda 8.1.1 SM

**Comment by tmetzke** [comment]
Came up here: https://camunda.slack.com/archives/C02JLRNQQ05/p1670934420969969

**Comment by chillleader** [comment]
Just reproduced it in SaaS, thanks for the detailed explanation!

I agree that this behavior is somewhat counterintuitive, but I'm not sure if this should rather be considered a feature request than a bug?

I think we can implement this in the next release, looks pretty straightforward.

**Comment by chDame** [comment]
Hi Pavel,

Great! I wasn't sure about opening it as a bug or feature request. But as it's not documented, I opted for the bug. Feel free to change it! Most important is to have it available soon :)

**Comment by tmetzke** [comment]
I'd opt for a feature request here. 
In general, the result expression is meant for mapping the results of the Connector to something else.
That does neither directly include nor exclude the possibility of using process variables in there.
But it wasn't our focus for this feature in the first iteration.

I am also not entirely sure if this is easy to support, actually. If I recall correctly, a Connector job only receives those variables defined in the `inputVariables` of its `@OutboundConnector` annotation, so you cannot easily hand-in all process variables.

**Comment by chDame** [comment]
Hi Tobias,

If I'm not mistaken, it's only called from:
https://github.com/camunda/connector-sdk/blob/d3aa9172bd9e00c256a14fcfbd90f713de0e78c5/runtime-util/src/main/java/io/camunda/connector/runtime/util/outbound/ConnectorJobHandler.java#L67

Where you could retrieve the process variables from the ActivatedJob ?

But I may be certainly be wrong, as I'm not familiar with that code :)

**Comment by tmetzke** [comment]
That is correct, but the `ActivatedJob` only contains those variables that were fetched by the Zeebe client. And it fetches the variables based on the `inputVariables` when it fetches the `ActivatedJob`.

Furthermore, since everyone can build a custom Connector runtime, this is also not guaranteed to work for Connectors in general. It might also not be possible in all environments to get all variables, because it might not be practical. For example, your Connectors could be run somewhere remote on GCP or AWS as a function that you'd want to call with as little data as possible, so you'd only send the necessary variables but nothing more.

Overall, I am not sure this is something we can/should easily build given the aspired environment-agnostic vision. We should maybe discuss how you could achieve something like the proposed with the means already there.

**Comment by nikku** [comment]
This came up once again via [this conversation](https://camunda.slack.com/archives/C02JLRNQQ05/p1673946835096149?thread_ts=1673887887.200989&cid=C02JLRNQQ05).

**Comment by nikku** [comment]
@chillleader You are assigned to this one. Is this something you plan to investigate?

Would be nice if we could understand possible solutions to this topic with the purpose to plan a potential fix. 

Would also be interesting to investigate existing guarantees that the engine gives you with [output mappings](https://docs.camunda.io/docs/next/components/concepts/variables/#output-mappings). How do these guarantees change once we move to [no-magic input mappings](https://github.com/camunda/team-connectors/issues/267).

**Comment by nikku** [comment]
I could imagine we could help the user tooling wise here (indicate that the variable that is available in the process is NOT available in the result mapping. However that would not give them a solution to their problem ("I want to access additional process context").

**Comment by chillleader** [comment]
To summarize: we **should not** commit to making all process variables available to the connector function, because it contradicts our runtime-agnostic approach

However, I can propose the following solution:
- Currently, **only connector result variables** are taken into account while evaluating the result expression
- Additionally, we can **merge** the connector result variables with the input variables. I think it's fair to guarantee that no matter which connector runtime is used, input mappings will be available to the runtime for every outbound connector.

To illustrate, let's imagine a user calls REST Connector and explicitly specifies that the following variables are passed to it as inputs:
```
{
  "comments": [
    {"author": "chillleader", "text": "hello"}
  ]
}
```
The user also specifies the following error expression:
```
comments: append(comments, {"author": response.body.author, "text": response.body.text)
```
REST Connector produces the following output:
```
{
  "request": {
    "body": {
      "author": "chillleader",
      "text": "hello again"
    }
  }
}
```
Then, Runtime merges the input variables with the connector result. The context for error expression evaluation therefore looks like this:
```
{
  "request": {
    "body": {
      "author": "chillleader",
      "text": "hello again"
    }
  },
  "comments": [
    {"author": "chillleader", "text": "hello"}
  ]
}
```
The provided error expression can now be evaluated correctly.

Now, there is a potential problem with input variables and connector results having the same name, but we can solve it by simply giving the connector results a priority and documenting it clearly.

In my opinion, this would work great together with no magic input mappings. @nikku WDYT?

**Comment by nikku** [comment]
> Now, there is a potential problem with input variables and connector results having the same name, but we can solve it by simply giving the connector results a priority and documenting it clearly.

:+1: 

> In my opinion, this would work great together with no magic input mappings. @nikku WDYT?

Sounds great indeed.

We could provide another (optional) input that allows users to pass additional process data into the connector:

```
// makes `someVariableIOnlyUseInResult available in connector context
additionalInput = {
  someVariableIOnlyUseInResult: processVariable
}

// uses variable being made available ^^
resultExpression = {
  foo: connectorResult + someVariableIOnlyUseInResult 
}
```

Thinking in terms of user assistance I could imagine that we validate the result expression against data available in the connector scope, thus identifying that `someVariableIOnlyUseInResult` is only there if mapped explicitly.

**Comment by nikku** [comment]
I still don't like that users would need to [explicitly map additional inputs](https://github.com/camunda/connector-sdk/issues/312#issuecomment-1405077127). 

One alternative I was considering: Could we move output mapping to `zeebe` core? The whole reason this currently is a limitation is that zeebe does not allow me to create output context, but only single output variables:

```xml
    <bpmn:serviceTask id="SERVICE_TASK">
      <bpmn:extensionElements>
        <zeebe:ioMapping>
          <zeebe:output source="=1" target="variable1" />
        </zeebe:ioMapping>
      </bpmn:extensionElements>
    </bpmn:serviceTask>
```

If we could map a set of outputs on the Zeebe side of things (where process variables are available) then we'd be able to support this _at the core_, without any connector hacks:

```xml
    <bpmn:serviceTask id="SERVICE_TASK">
      <bpmn:extensionElements>
        <zeebe:ioMapping>
          <zeebe:outputExpression expression="={ variable1: variable1 + someProcessVariable }" />
        </zeebe:ioMapping>
      </bpmn:extensionElements>
    </bpmn:serviceTask>
```

That solution would violate [no magic input mappings](https://github.com/camunda/team-connectors/issues/267). 

But solutions exists to [circumvent that, too](https://github.com/camunda/zeebe/issues/11163#issuecomment-1405119992).

**Comment by chillleader** [comment]
> We could provide another (optional) input that allows users to pass additional process data into the connector. Otherwise we'd still have the issue that users cannot map arbitary data into the connector:

Yes, a separate field will apparently be necessary. It adds one more bit into the complexity pile for the connectors user - which I also don't like that much.

Also, this approach will require additional support in the Modeler, right? Since there's no functionality in element templates currently that would allow to dynamically add input mappings.

> If we could map a set of outputs on the Zeebe side of things (where process variables are available) then we'd be able to support this at the core, without any connector hacks:

This actually sounds great. I would prefer this issue to be solved on the Zeebe side as well, if possible.

**Comment by bastiankoerber** [comment]
@sbuettner please discuss it tomorrow in the architecture sync

**Comment by sbuettner** [comment]
We discussed this topic in our architecture session and decided against an implementation for now. @tmetzke already mentioned the variable declaration that we would have to extend and open for all process variables.

We recommend to make use of a script task using a FEEL expression which has access to all process variables including the ones mapped based on the connector result.
